File: benchmarks/tsb/bench_dataframe_apply.ts
/**
 * Benchmark: dataframe_apply — apply a function across rows of a 10k-row DataFrame
 * (reduced size due to JS per-row overhead)
 */
import { DataFrame } from "../../src/index.js";

const ROWS = 10_000;
const WARMUP = 3;
const ITERATIONS = 10;

const a = Array.from({ length: ROWS }, (_, i) => i * 1.0);
const b = Array.from({ length: ROWS }, (_, i) => i * 2.0);
const df = DataFrame.fromColumns({ a, b });

for (let i = 0; i < WARMUP; i++) {
  df.apply((row) => (row.at("a") as number) + (row.at("b") as number), 1);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  df.apply((row) => (row.at("a") as number) + (row.at("b") as number), 1);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "dataframe_apply",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_dataframe_apply.py
"""Benchmark: dataframe_apply — apply a function across rows of a 10k-row DataFrame"""
import json, time
import numpy as np
import pandas as pd

ROWS = 10_000
WARMUP = 3
ITERATIONS = 10

a = np.arange(ROWS, dtype=np.float64)
b = np.arange(ROWS, dtype=np.float64) * 2.0
df = pd.DataFrame({"a": a, "b": b})

for _ in range(WARMUP):
    df.apply(lambda row: row["a"] + row["b"], axis=1)

start = time.perf_counter()
for _ in range(ITERATIONS):
    df.apply(lambda row: row["a"] + row["b"], axis=1)
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "dataframe_apply",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
